File: LeetCode/best-time-to-buy-and-sell-stock.cpp
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int max_profit_till_now = 0;
        int least_price = INT_MAX;
        int profit_today = 0;
        for (auto price : prices)
        {
            profit_today = price - least_price;
            least_price = min(least_price, price);
            max_profit_till_now = max(max_profit_till_now, profit_today);
        }
        return max_profit_till_now;
    }
};

File: LeetCode/best-time-to-buy-and-sell-stock.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution {
  public int maxProfit(int[] prices) {
    int maxProfit = 0;
    int profitToday = 0;
    int leastPrice = Integer.MAX_VALUE;
    for (int price : prices) {
      profitToday = price - leastPrice;
      leastPrice = Math.min(leastPrice, price);
      maxProfit = Math.max(maxProfit, profitToday);
    }
    return maxProfit;
  }
}

class Solution {
  public int maxProfit(int[] prices) {
    // We want to keep tracking the smallest value that we see,
    // since we would probably want to buy at that price.
    int minPrice = Integer.MAX_VALUE;
    int maxProfit = Integer.MIN_VALUE;

    // go through all the prices
    for (int price : prices) {
      // Is this the lowest price we have found so far?
      if (price < minPrice) {
        minPrice = price;
      } else if (price - minPrice > maxProfit) {
        // is this the highest selling price we have found so far?
        maxProfit = price - minPrice;
      }
    }
    return maxProfit;
  }
}